### PR TITLE
Don't use setState for disabled KeyboardAvoidingView to avoid re-renders

### PR DESCRIPTION
**Pull request by adamgrzybowski**
## Summary:

There are two reasons to apply these changes:
- We don't need to re-render the `KeyboardAvoidingView` if it is disabled. It may be especially useful in combination with [react-navigation](https://reactnavigation.org/) where we could disable `KeyboardAvoidingView` for screens that are not focused 
- They fix the problem with the `KeyboardAvoidingView` wrapped inside the [react-freeze](https://github.com/software-mansion/react-freeze) component. Similarly, as above, it is useful when we want to freeze screens that are not visible for the user.

## Changelog:
[GENERAL] [CHANGED] Don't use setState for disabled KeyboardAvoidingView to avoid re-renders

## Test Plan:

- Check if the KeyboardAvoidingView works as expected.

### CODE REVIEW

**Comment by blakef** [comment]
Looks like you've got flow test failing, can you take a look?

File: packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js [hunk 1]
```diff
@@ -129,20 +130,32 @@ class KeyboardAvoidingView extends React.Component<Props, State> {
     }
   };
 
+  // Avoid unnecessary renders if the KeyboardAvoidingView is disabled.
+  _setBottom = (value: number) => {
+    const {enabled = true} = this.props;
+    this._bottom = value;
+    if (enabled === true) {
```
Comment (Pranav-yadav) [inline packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js:137]:
```suggestion
    if (enabled) {
```
Self explanatory, though stylistic :)

File: packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js [hunk 2]
```diff
@@ -129,20 +130,32 @@ class KeyboardAvoidingView extends React.Component<Props, State> {
...
-    if (this.state.bottom === height) {
+    if (this._bottom === height) {
       return;
     }
 
-    if (duration && easing) {
+    this._setBottom(height);
+
+    const {enabled = true} = this.props;
+    if (enabled === true && duration && easing) {
```
Comment (Pranav-yadav) [inline packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js:158]:
```suggestion
    if (enabled && duration && easing) {
```
Self explanatory, less verbose :)

File: packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js [hunk 3]
```diff
@@ -152,9 +165,15 @@ class KeyboardAvoidingView extends React.Component<Props, State> {
         },
       });
     }
-    this.setState({bottom: height});
   };
 
+  componentDidUpdate(_: Props, prevState: State): void {
+    const {enabled = true} = this.props;
+    if (enabled === true && this._bottom !== prevState.bottom) {
```
Comment (Pranav-yadav) [inline packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js:172]:
```suggestion
    if (enabled && this._bottom !== prevState.bottom) {
```

File: packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js (hunk 1 above)
Comment (blakef) [reply packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js:137]: @adamgrzybowski  I see you made that change on your 2nd commit.  What was your thinking?

Comment (adamgrzybowski) [reply packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js]: @blakef @Pranav-yadav done

**Comment by adamgrzybowski** [comment]
Is there anything else I should answer? @blakef @Pranav-yadav

**Comment by efstathiosntonas** [comment]
Hi, what's the status on this PR? Is it ready to be merged?

**Comment by kelset** [comment]
given that checks are all passing, @cortinico maybe you can help find someone from Meta to get this in?

**Comment by toscalivia83** [comment]
Hello @adamgrzybowski,
I'm having an error in my RN 0.73.2 project, only in production on iOS: 
on press on an Input, after displaying the keyboard, my app freezes.

I upgraded all the dependencies and it did not fix it.

No error is sent on focus to Sentry although I asked for one to be sent onFocus and onChangeText. I can't add anything with the keyboard. Although, in some cases, I can type something with the keyboard but it freezes 2 or 3 seconds later... I can't see any difference between when it works and when it does not.

This is my package.json, could you help me debug this? 🙏 

```
{
  "name": "XXXX",
  "version": "2.3.2",
  "Android": {
    "versionCode": 57,
    "versionName": "2.3.2"
  },
  "private": true,
  "main": "index.js",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint .",
    "android:dev": "ENV=development ./envscript.sh && ENVFILE=.env.development && react-native run-android --mode=developmentdebug",
    "android:dev-release": "ENV=development ./envscript.sh && ENVFILE=.env.development && react-native run-android --mode=developmentrelease",
    "android:prod": "ENV=production ./envscript.sh && ENVFILE=.env.production && react-native run-android --mode=productiondebug",
    "android:prod-release": "ENV=production ./envscript.sh && ENVFILE=.env.production && react-native run-android --mode=productionrelease",
    "android:bundle-release": "ENV=production ./envscript.sh && cd android && npx react-native run-android --mode=ProductionRelease && ENVFILE=.env.production ./gradlew bundleRelease",
    "android:assemble-release": "ENV=production ./envscript.sh && cd android && ENVFILE=.env.production ./gradlew assembleRelease",
    "ios:dev": "ENV=development ./envscript.sh && ENVFILE=.env.development react-native run-ios",
    "ios:prod-release-on-my-device": "ENV=production ./envscript.sh && ENVFILE=.env.production react-native run-ios --device \"Marion's iPhone\""
  },
  "dependencies": {
    "@babel/runtime": "^7.23.8",
    "@bam.tech/react-native-image-resizer": "^3.0.7",
    "@notifee/react-native": "^7.8.2",
    "@react-native-async-storage/async-storage": "^1.15.9",
    "@react-native-clipboard/clipboard": "^1.11.2",
    "@react-native-community/datetimepicker": "6.7.3",
    "@react-native-firebase/analytics": "^18.7.3",
    "@react-native-firebase/app": "^18.7.3",
    "@react-native-firebase/auth": "^18.7.3",
    "@react-native-firebase/firestore": "^18.7.3",
    "@react-native-firebase/functions": "^18.7.3",
    "@react-native-firebase/messaging": "^18.7.3",
    "@react-native-firebase/storage": "^18.7.3",
    "@react-native-picker/picker": "^2.6.1",
    "@react-navigation/bottom-tabs": "^6.5.11",
    "@react-navigation/native": "^6.1.9",
    "@react-navigation/stack": "^6.3.20",
    "@stripe/stripe-react-native": "^0.35.1",
    "firebase": "^10.7.1",
    "react": "18.2.0",
    "react-native": "0.73.2",
    "react-native-app-intro-slider": "^4.0.4",
    "react-native-calendars": "^1.1284.0",
    "react-native-chart-kit": "^6.12.0",
    "react-native-config": "^1.4.5",
    "react-native-document-picker": "^8.1.3",
    "react-native-gesture-handler": "^1.10.3",
    "react-native-google-places-autocomplete": "^2.2.0",
    "react-native-image-crop-picker": "^0.40.2",
    "react-native-screens": "^3.29.0",
    "react-native-splash-screen": "^3.3.0",
    "react-native-vector-icons": "^10.0.3"
  },
  "devDependencies": {
    "@babel/core": "^7.20.0",
    "@babel/preset-env": "^7.20.0",
    "@react-native-picker/picker": "^2.2.1",
    "@react-native/babel-preset": "0.73.19",
    "@react-native/eslint-config": "0.73.2",
    "@react-native/metro-config": "0.73.3",
    "@react-native/typescript-config": "0.73.1",
    "@sentry/react-native": "^4.15.2",
    "@tanstack/query-async-storage-persister": "^4.29.5",
    "@tanstack/react-query": "^4.29.5",
    "@tanstack/react-query-persist-client": "^4.29.5",
    "@types/react": "^18.2.6",
    "@types/react-test-renderer": "^18.0.0",
    "AsyncStorage": "^0.1.5",
    "babel-jest": "^29.6.3",
    "dayjs": "^1.11.8",
    "eslint": "^8.19.0",
    "i18next": "^21.3.3",
    "jest": "^29.6.3",
    "moment": "^2.29.1",
    "numeral": "^2.0.6",
    "prettier": "2.8.8",
    "qs": "^6.10.4",
    "react-i18next": "^11.13.0",
    "react-native-image-crop-picker": "^0.39.0",
    "react-native-image-viewing": "^0.2.2",
    "react-native-localize": "^2.1.5",
    "react-native-maps": "^0.27.1",
    "react-native-modal-datetime-picker": "^13.1.2",
    "react-native-push-notification": "^8.1.1",
    "react-native-safe-area": "^0.5.1",
    "react-native-safe-area-context": "^3.3.2",
    "react-native-svg": "^13.9.0",
    "react-native-swipe-gestures": "^1.0.5",
    "react-native-web": "0.19.4",
    "react-native-webview": "^13.6.2",
    "react-test-renderer": "18.2.0",
    "typescript": "5.0.4"
  },
  "engines": {
    "node": ">=18"
  }
}
```

**Pull request closed without merging**